README.md

# 项目开发指导

## 技术栈

- iuap design
- nodejs
- koa
- gulp
- director
- less
- require
- babel
- git
- react

## iUAP Design相关的学习资料

请直接前往我们的[官网](http://design.yyuap.com).

## 开发准备

### Nodejs相关

- 全局安装nodejs环境，下载请前往[nodejs.org](https://nodejs.org)
- 如果你还不太了解Nodejs，请前往[Node.js知识详解](http://guoyongfeng.github.io/idoc/html/%E6%8A%80%E6%9C%AF%E5%88%86%E4%BA%AB/Node.js%E7%9F%A5%E8%AF%86%E8%AF%A6%E8%A7%A3.html)。
- [Nodejs环境配置](http://guoyongfeng.github.io/idoc/html/%E6%8A%80%E6%9C%AF%E5%88%86%E4%BA%AB/Nodejs%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE.html)
- [NPM的使用方法](http://guoyongfeng.github.io/idoc/html/%E6%8A%80%E6%9C%AF%E5%88%86%E4%BA%AB/NPM%E7%9A%84%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95.html)

### Git相关

- 下载Git客户端，方便命令行操作，下载请前往[git-scm.com](https://git-scm.com/download/)
- 如果你还不熟悉基本的git命令行操作，请前往学习[Git和Github快速入门指南](http://guoyongfeng.github.io/idoc/html/%E6%8A%80%E6%9C%AF%E5%88%86%E4%BA%AB/Git%E5%92%8CGithub%E5%BF%AB%E9%80%9F%E5%85%A5%E9%97%A8%E6%8C%87%E5%8D%97.html)。

### 编辑器和浏览器

- 选一款自己喜欢的编辑器：Atom/Sublime Text 3/Webstorm，贴一下集中的导航[下载地址](http://guoyongfeng.github.io/idoc/html/%E5%9B%A2%E9%98%9F/%E6%96%B0%E5%91%98%E5%B7%A5%E5%85%A5%E8%81%8C%E6%8C%87%E5%8D%97.html)。
- 推荐下载Chrome或Firefox进行前端开发调试


## 下载项目并且启动

```
$ git clone git@github.com:iuap-design/cloud-starter-kit.git
$ cd fe
# 切换到develop分支进行开发
$ git checkout develop
# 安装依赖
$ npm install
# 启动开发
$ npm run dev
# 访问http://localhost:8080
```

## 开发和调试

两条命令支持开发：
- 启动开发
```
$ npm run dev
```

- 清空dist目录下的资源
```
$ npm run clean
```
### 目录说明

- docs/ 文档
- examples/ 示例
- gulpfile.js gulp构建脚本
- dist/ 编译产出的资源目录
- index.html 调试首页
- mock.config.js 模拟数据配置
- mockData/ 模拟接口json数据
- node_modules/ node第三方包
- package.json npm管理文件
- README.md 项目说明文件
- require.config.js requirejs配置
- server/ 本地服务和代理配置
- src/ 开发源代码目录
  - pages
		* time
		* home
		* ssc
		* corehr
  - static 图片、音频、视频...
  - util 公共方法或是插件的抽取
  - styles 全局样式文件
- vendor/ 第三方插件、框架、类库、路由等


### 开发起步

- 在 `src/pages` 目录下新增功能页面，可参考其他目录的代码。
- 为了方便调试，在 `index.html` 文件内新增你的页面 URL，具体可以看里面的注释部分。

```
$ cd src/pages/time && mkdir demo
$ cd demo && touch demo.html demo.js demo.css
```

代码清单：`src/pages/time/demo/demo.html`

```
.demo {
  border: 1px solid #ccc;
}
```

代码清单：`src/pages/time/demo/demo.css`

```
<div class="demo">
  <h3>
    这是一个示例页面的html代码片段
  </h3>
</div>
```
代码清单：`src/pages/time/demo/demo.js`

```
define(function(require, module, exports){
  // 引入相关的功能插件或模块
  var html = require('text!./demo.html');
  var demoStyles = require('css!./demo.css');

  function DemoLogic(){
    this.init = function(){
      alert('hey, man!');
    }
  }

  // 以下部分代码基本是格式化的，暴露的init方法用于和portal进行集成
  return {
		init: function( content ) {
			// 插入html片段
			content.innerHTML = html;
			// 执行我们的业务主逻辑
			new DemoLogic().init();
		}
	}

})

```

对以上代码的基本说明：引入了相应的html片段和css样式，另外，由于是和portal集成，在portal部分加载了以下公共的类库或是框架：

1. `bootstrap`
2. `director` **处理前端路由**
3. `jquery`
4. `knockout` **前端MVVM框架**
5. `require`
6. `hrfonts`  **hrfonts是UE通过iconfont制作的图标字体库**
7. `font-awesome`
8. `uui`

> 其中，UUI是iUAP前端技术部提供的技术产品集合，包含iuap-design、基于ko的model增强库datatable、各种插件

这样基本编辑之后，在`index.html`文件中按照注释说明，添加你新增页面的导航，代码如下：

```
<a href="#/dist/pages/demo/demo">
  <i class="demo-menu-icon u-color-grey-400 fa fa-home fa-2x" role="presentation"></i>
  示例
</a>
```

### 数据模拟和联调

调试是开发过程中关键而重要的一步，经常在这个阶段出现一些问题，建议的开发步骤是：
1. 前端本地模拟数据调试，跑通业务逻辑
2. 前后端本地调试，跑通接口数据
3. 部署到服务器上联调

#### 前端本地模拟数据调试

- 数据模拟配置文件： `mock.config.js`，先在这个文件中新增一条模拟路由：

```
{
	  type:"get",         // 请求类型
	  url:"/demo/data",   //请求url
	  json:"demo.json"    //模拟数据文件，路径地址相对于mockData目录，**可以创建二级甚至三级目录，不要只在一级目录创建文件**
}
```

- 在mockData目录下新增相应的json模拟数据

```
$ cd mockData && touch demo.json
```

编辑demo.json里面的模拟数据，代码清单：`mockData\demo.json`

```
{
    "statusCode": 200,
    "message": "操作成功",
    "method": null,
    "data": {
        "name": "guoyongfeng"
    },
    "callbackType": null
}
```

具体应该写什么样的模拟数据，需要前后端提前约定模拟接口和接口数据。

- 现在我们是本地模拟数据，请修改`gulpfile.js`里面的`useProxy`为`false`。
- 在demo.js文件里面编辑代码，获取我们模拟的数据

```
define(function(require, module, exports){

  // 引入相关的功能插件或模块
  var html = require('text!./demo.html');
  var demoStyles = require('css!./demo.css');

  function DemoLogic(){
    this.init = function(){
      alert('hey, man!');
      this.getData();
    }
  }

  DemoLogic.prototype = {
    getData: function(){
      $.ajax({
        url: '/demo/data',
        type: 'get',
        dataType: 'json',
        success: function( res ){
          alert( res.data.name );
        }
      })
    }
  };

  // 以下部分代码基本是格式化的，暴露的init方法用于和portal进行集成
  return {
		init: function( content ) {
			// 插入html片段
			content.innerHTML = html;
			// 执行我们的业务主逻辑
			new DemoLogic().init();
		}
	}

})

```

#### 前后端本地联调

- 将`app.js`文件中的`useProxy`值设置为`true`即可
- `app.js`里有个host字段用于配置请求哪个服务器地址的数据，默认是`http://172.20.14.92`，在前后端本地联调过程中，可以修改这个host指到对应的服务端工程师本机的服务地址，如：`var host = 'http://10.1.255.252:8081';`


CHANGELOG.md
# 1.1.0（2016-05-03）

> 本次升级以实现部分前端基础功能为主

## 1. CSS 模块化方案
- [ ] [CSS Module -- 使用webpack的css-loader可以解决这个问题，通过编译期 renaming 的方式为 CSS 引入了局部变量]()

## 2. 几种业务数据的前端展现

### 数据下拉展示
- [ ] [简单数据的select]()
- [ ] [autocomplete + Solr的智能匹配搜索下拉]()

### 弹框内的结构化数据搜索及校验
- [ ] [带分页的表数据]()
- [ ] [tree树形结构数据]()
- [ ] [左树右表数据]()
- [ ] [日期（年月、年月日、时分秒、年月日时分秒）数据]()

## 3. 面包屑导航和相关操作链接

- [ ] [单独的面包屑导航，和portal的导航关联]()
- [ ] [每个导航的右侧都会有业务功能的相关操作链接]()

## 4. 高级搜索
- [ ] [搜索的选项、样式、可扩展性]()

## 5. 全站UE样式升级
- [ ] [抓住这次机会统一样式文件，缩减大部分页面的CSS文件]()
- [ ] [portal部分封装底色为#eeeff1，部分为纯白色背景]()
- [ ] [白色区块和底色相间]()
- [ ] [图标字体加载，避免加载多套图标资源]()
- [ ] [portal和业务页面的色块关联和区分]()

# 1.0.1-rc1（2016-04-19）

## 1.增加gulp的task任务
- [x] [代码压缩]()
- [ ] [给文件添加MD5]()

## 2.规范目录命名和文件命名
- [x] [业务目录命名简洁，复杂命名用"-"连接]()
- [x] [文件统一命名为index开头，编辑等页面除外。`备注：现有页面的文件名和目录名称暂时不作修改，新增的业务功能页面统一按照上面两条执行`]()

## 3.公共样式抽取

- [x] [抽取页面公共样式，styles目录，并统一在common.css文件中import]()
- [ ] [干掉80%页面的CSS文件]()

## 4.代码规范的落实
- [ ] [**重要：优化业务代码组织方式（写法放在example目录）**]()
- [x] [代码注释]()
- [ ] [缩进统一为4个空格]()
- [ ] ["="号两边空格，定义的变量和逻辑之间空行]()
- [ ] [JS变量命名遵循小驼峰，CSS样式变量命名用"-"连接]()
- [ ] [优化逻辑写法]()


## 5.公共方法抽取
- [x] [公共方法抽离到util目录下，并且增加使用文档]()
- [ ] [二个页面以上公用的方法抽取]()
- [ ] [一个页面中重复用到的逻辑抽取]()

## 6.业务插件库建设
- [ ] [功能插件的使用整理及示例]()
- [ ] [待开发插件计划]()

# 1.0.0（2016-04-11）

## 1.重新调整目录
- [x] [src为源码]()
- [x] [产出的hrcloud目录下资源为和portal集成的资源]()
- [x] [独立vendor目录和src同级]()


## 2.使用Gulp构建工具进行前端资源构建
- [x] [编译]()
- [ ] [打包合并]()
- [ ] [压缩]()
- [x] [校验]()
- [ ] [MD5]()


## 3.CSS样式
- [x] [支持LESS预处理器编写CSS样式文件]()

## 4.支持ES6
- [x] [可以在项目中使用ES6的新特性]()
- [x] [产出AMD的JS文件与portal集成]()


gulpfile.js
var gulp = require('gulp');
var less = require('gulp-less');
var clean = require('gulp-clean');
var babel = require('gulp-babel');
var copy = require('gulp-copy');
var concat = require('gulp-concat');
var rename = require('gulp-rename');
var util = require('gulp-util');
var minifycss = require('gulp-minify-css');
var uglify = require('gulp-uglify');
var koa = require('koa');

var app = koa();
var MockServer = require('./server/mockserver');

var proxyServerConfig = {
    // 是否启用后端代理
    //true -> 使用服务器数据，false -> 使用本地模拟数据
    useProxy : true,
    proxyIgnore:[
        "/",
        "/css.js",
        "/text.js",
        "/src/*",
        "/vendor/*",
        "/dist/*",
        "/conf/*"
    ],
    host : 'http://localhost:8081',
    username : '',
    password : '',
    context : '',
    serverport : 8080
};

/**
 * 公共错误处理函数
 * 使用示例：
 *  .pipe(uglify())
 .on('error', errHandle)
 */
function errHandle(err) {
    console.log(err);
    util.log(err.fileName + '文件编译出错，出错行数为' + err.lineNumber + '，具体错误信息为：' + err.message);
    this.end();
};

// 编译 src 下所有的 html 文件到 dist 目录
gulp.task('html', function(){
    gulp.src('src/**/**/*.html')
        .pipe(rename(function(path){
            path.dirname += '';
        }))
        .pipe(gulp.dest("./dist"));
})

// 编译 src 下所有的 xml 文件到 dist 目录
gulp.task('xml', function(){
    gulp.src('src/widget/**/*.xml')
        .pipe(rename(function(path){
            path.dirname += '';
        }))
        .pipe(gulp.dest("./dist/widget"));
})
// copy 业务目录下的图片到相应的目录
gulp.task('copy:img', function(){
    gulp.src(['src/**/**/**/*.png', 'src/**/**/**/*.jpg'])
        .pipe(rename(function(path){
            path.dirname += '';
        }))
        .pipe(gulp.dest("./dist"));
});

// 完整 copy vendor 目录下的资源到 dist
gulp.task('copy:vendor', function(){
    gulp.src('./vendor/**')
        .pipe(copy('./dist'));
});

// copy static目录下的资源到 dist
gulp.task('copy:static', function(){
    gulp.src('src/static/**')
        .pipe(gulp.dest('dist/static'))
});

// 匹配所有 css 进行 less 编译
gulp.task('less', function(){
    gulp.src('src/**/*.css')
        .pipe(less())
        .pipe(gulp.dest('dist'));
});

//匹配 css 通过 flexible 编译成rem单位
// gulp.task('flexible', function() {
//     var processors = [px2rem({remUnit: 10})];
//     return gulp.src('src/pages/corehr/organiazation/index.css')
//         .pipe(postcss(processors))
//         .pipe(gulp.dest('dist'));
// });

gulp.task('less:dist', function(){
    gulp.src(['src/**/*.css', '!src/styles/*.css'])
        .pipe(less())
        .pipe(minifycss())
        .pipe(gulp.dest('dist'));
});

gulp.task('common', function(){
    gulp.src(['src/styles/*.css', '!src/styles/common.css'])
        .pipe(less())
        .pipe(concat('common.css'))
        .pipe(minifycss())
        .pipe(gulp.dest('dist/styles'));
});

// 编译 JS 代码，支持 ES6 语法编译
gulp.task('es2015',function(){
    gulp.src(['src/**/*.js'])
        .pipe(babel())
        .on('error', errHandle)
        .pipe(gulp.dest('dist'));
});

gulp.task('es2015:dist',function(){
    gulp.src(['src/**/*.js'])
        .pipe(babel())
        .pipe(uglify())
        .on('error', errHandle)
        .pipe(gulp.dest('dist'));
});

// 监听文件改动，执行相应任务
gulp.task('watch', function(){
    gulp.watch('src/**/**/*.css',  ['less']);
    gulp.watch('src/**/*.js',  ['es2015']);
    gulp.watch('src/**/**/*.html', ['html']);
});

// 清空 dist 目录下的资源
gulp.task('clean',function() {
    gulp.src('dist/*', {read: false})
        .pipe(clean({force: true}));
});

// start mock server
gulp.task('mock-server', function(){
    proxyServerConfig.app = app;
    var mockServer = new MockServer(proxyServerConfig);
    mockServer.start(proxyServerConfig);
});

gulp.task('copy', ['copy:static', 'copy:img', 'copy:vendor']);
gulp.task('before', ['copy', 'html','xml', 'less','es2015','common']);
gulp.task('prod', ['copy', 'html','xml', 'less:dist', 'es2015','common']);
gulp.task('default', ['before', 'mock-server', 'watch']);
